sdk/message/src/legacy.rs

```diff
@@ -724,11 +726,11 @@ impl Message {
         false
     }
 
-    /// Returns `true` if any account is the BPF upgradeable loader.
-    pub fn is_upgradeable_loader_present(&self) -> bool {
-        self.account_keys
-            .iter()
-            .any(|&key| key == bpf_loader_upgradeable::id())
+    /// Inspect all message keys for loader v3 or v4
+    pub fn is_upgradeable_loader_present(&self, enable_loader_v4: bool) -> bool {
+        self.account_keys.iter().any(|key| {
+            bpf_loader_upgradeable::check_id(key) || (enable_loader_v4 && loader_v4::check_id(key))
+        })
```

Comment:
we'd need a similar piping of this variable to the transaction view; see `ResolvedTransactionView::cache_is_writable` and `test_demote_writable_program` in the same file.